File: src/codegen/indent.py
from __future__ import annotations


def apply_indent(text: str, base_indent: str, auto_indent: bool) -> str:
    """Prepend *base_indent* to every non-empty line in *text* (§5.1).

    If *auto_indent* is False, return *text* unchanged.
    The final newline (if any) is preserved exactly as-is.
    """
    if not auto_indent or not base_indent:
        return text

    ends_with_newline = text.endswith("\n")
    lines = text.splitlines()

    result = []
    for line in lines:
        if line:  # non-empty: add indent
            result.append(base_indent + line)
        else:     # blank: leave empty (no trailing whitespace)
            result.append("")

    joined = "\n".join(result)
    if ends_with_newline:
        joined += "\n"
    return joined

File: cpp/src/indent.cpp
#include "indent.hpp"
#include "utils.hpp"

std::string apply_indent(const std::string& text,
                         const std::string& base_indent,
                         bool auto_indent) {
    if (!auto_indent || base_indent.empty())
        return text;

    bool ends_nl = ends_with(text, "\n");
    auto lines = splitlines(text, false);

    std::string result;
    result.reserve(text.size() + lines.size() * base_indent.size());
    for (size_t i = 0; i < lines.size(); i++) {
        if (!lines[i].empty())
            result += base_indent + lines[i];
        // blank lines get no trailing whitespace
        if (i + 1 < lines.size())
            result += '\n';
    }
    if (ends_nl)
        result += '\n';
    return result;
}
